### PR TITLE
Added Appointment Slip Download Functionality

### PR DESCRIPTION
**Pull request by dhhruv**
Testing Needed...

### CODE REVIEW

**Comment by dhhruv** [comment]
In response to #385 .

**Comment by bombardier-gif** [comment]
Neat ! let us test this out for a while.

**Comment by Nakul93** [comment]
@dhhruv So I've tried out this PR but it didn't download the PDF file. Does it try to store it in the main directory?

File: src/utils.py [hunk 1]
```diff
@@ -580,6 +583,21 @@ def book_appointment(request_header, details, mobile, generate_captcha_pref='n')
...
+                    appSlipBase = (
+                        APPOINTMENT_SLIP_URL
+                        + f"&appointment_id={resp.json()['appointment_id']}"
+                    )
+                    appslip = requests.get(appSlipBase, headers=request_header)
+                    with open(
+                        f"{resp.json()['appointment_id']}.pdf", "wb"
+                    ) as appSlipPdf:
+                        appSlipPdf.write(appslip.content)
+
```
Comment (Nakul93) [inline src/utils.py:597]: Could you add a print statement for debug purposes and to let the user know that if the PDF was downloaded or not

File: src/utils.py [hunk 2]
```diff
@@ -580,6 +583,21 @@ def book_appointment(request_header, details, mobile, generate_captcha_pref='n')
...
+                except Exception as e:
+                    pass
+
```
Comment (Nakul93) [inline src/utils.py:600]: Put a print stmt here as well

File: src/utils.py (hunk 1 above)
Comment (shailesh) [reply src/utils.py:597]:
```
if os.path.exists(f"{resp.json()['appointment_id']}.pdf"):
    print("\nDownload Successful. Check the Current Working Directory for the Appointment Slip.")
else:
    print("\nAppointment Slip Download Failed...")
```

Comment (shailesh) [reply src/utils.py:597]: Als add a check for status 200/400 
Would be better to Name file as : [**mobile_no__appoint_slip_appointment_id**]

**Comment by dhhruv** [comment]
> @dhhruv So I've tried out this PR but it didn't download the PDF file. Does it try to store it in the main directory?

It should download in the current working directory.

**Review by shailesh — CHANGES_REQUESTED** [review]

**Comment by dhhruv** [comment]
```
try:
        appSlipBase = (
            APPOINTMENT_SLIP_URL
            + f"&appointment_id={resp.json()['appointment_confirmation_no']}"
        )
        appslip = requests.get(appSlipBase, headers=request_header)
        with open(
            f"{resp.json()['appointment_confirmation_no']}.pdf", "wb"
        ) as appSlipPdf:
            appSlipPdf.write(appslip.content)
        if os.path.exists(
            f"{resp.json()['appointment_confirmation_no']}.pdf"
        ):
            print(
                "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
            )
        else:
            print(f"{Fore.RED}", end="")
            print("\nAppointment Slip Download Failed...")
        
        except Exception as e:
        print(f"{Fore.RED}", end="")
        print(str(e))
        print(f"{Fore.RESET}", end="")
        print("\n\n")
        beep(WARNING_BEEP_DURATION[0], WARNING_BEEP_DURATION[1])
```
@Nakul93 Try this one. Currently I don't have any open accounts to check, but this must work and I've added changes you asked.

File: src/utils.py (hunk 1 above)
Comment (Nakul93) [reply src/utils.py:597]: +1 @shailesh. Having the file name as mobile_no.pdf will make more sense.

**Comment by Nakul93** [comment]
@dhhruv Got this error while opening the pdf file
image

**Comment by dhhruv** [comment]
> @dhhruv Got this error while opening the pdf file
> image

Can you decode and print the incoming data so that we can know what's actually coming through @Nakul93 ?

**Comment by Nakul93** [comment]
> > @dhhruv Got this error while opening the pdf file
> > image
> 
> Can you decode and print the incoming data so that we can know what's actually coming through @Nakul93 ?

How to do that?

**Comment by dhhruv** [comment]
```
try:
                    appSlipBase = (
                        APPOINTMENT_SLIP_URL
                        + f"?appointment_id={resp.json()['appointment_confirmation_no']}"
                    )
                    appslip = requests.get(appSlipBase, headers=request_header)
                    with open(
                        f"{resp.json()['appointment_confirmation_no']}.pdf", "wb"
                    ) as appSlipPdf:
                        appSlipPdf.write(appslip.content)
                    if os.path.exists(
                        f"{resp.json()['appointment_confirmation_no']}.pdf"
                    ):
                        print(
                            "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
                        )
                    else:
                        print("\nAppointment Slip Download Failed...")

                except Exception as e:
                    print(str(e))
                    print("\n\n")
```
**OR**
```
try:
                    appSlipBase = (
                        APPOINTMENT_SLIP_URL
                        + f"?appointment_id={resp.json()['appointment_id']}"
                    )
                    appslip = requests.get(appSlipBase, headers=request_header)
                    with open(
                        f"{resp.json()['appointment_id']}.pdf", "wb"
                    ) as appSlipPdf:
                        appSlipPdf.write(appslip.content)
                    if os.path.exists(
                        f"{resp.json()['appointment_id']}.pdf"
                    ):
                        print(
                            "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
                        )
                    else:
                        print("\nAppointment Slip Download Failed...")

                except Exception as e:
                    print(str(e))
                    print("\n\n")
```
There was a minor mistake, I've taken care of that. But there's a bit of a confusion as in the Protected API (CoWIN) sends appointment id in json as `appointment_id` but in real it is sending as `appointment_confirmation_no` so I've written 2 code snippets and one of them should definitely work. Can you please take a look @Nakul93 @shailesh and check/test which one is working?

**Comment by dhhruv** [comment]
> > > @dhhruv Got this error while opening the pdf file
> > > image
> > 
> > 
> > Can you decode and print the incoming data so that we can know what's actually coming through @Nakul93 ?
> 
> How to do that?

No needs, the error is solved and any one of the two code snippets shared above will work. Please test and give feedback.

**Comment by Nakul93** [comment]
> ```
> try:
>                     appSlipBase = (
>                         APPOINTMENT_SLIP_URL
>                         + f"?appointment_id={resp.json()['appointment_confirmation_no']}"
>                     )
>                     appslip = requests.get(appSlipBase, headers=request_header)
>                     with open(
>                         f"{resp.json()['appointment_confirmation_no']}.pdf", "wb"
>                     ) as appSlipPdf:
>                         appSlipPdf.write(appslip.content)
>                     if os.path.exists(
>                         f"{resp.json()['appointment_confirmation_no']}.pdf"
>                     ):
>                         print(
>                             "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
>                         )
>                     else:
>                         print("\nAppointment Slip Download Failed...")
> 
>                 except Exception as e:
>                     print(str(e))
>                     print("\n\n")
> ```
> 

This is same as before right?

> **OR**
> 
> ```
> try:
>                     appSlipBase = (
>                         APPOINTMENT_SLIP_URL
>                         + f"?appointment_id={resp.json()['appointment_id']}"
>                     )
>                     appslip = requests.get(appSlipBase, headers=request_header)
>                     with open(
>                         f"{resp.json()['appointment_id']}.pdf", "wb"
>                     ) as appSlipPdf:
>                         appSlipPdf.write(appslip.content)
>                     if os.path.exists(
>                         f"{resp.json()['appointment_id']}.pdf"
>                     ):
>                         print(
>                             "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
>                         )
>                     else:
>                         print("\nAppointment Slip Download Failed...")
> 
>                 except Exception as e:
>                     print(str(e))
>                     print("\n\n")
> ```
> 

This doesn't work
image

**Comment by dhhruv** [comment]
> > ```
> > try:
> >                     appSlipBase = (
> >                         APPOINTMENT_SLIP_URL
> >                         + f"?appointment_id={resp.json()['appointment_confirmation_no']}"
> >                     )
> >                     appslip = requests.get(appSlipBase, headers=request_header)
> >                     with open(
> >                         f"{resp.json()['appointment_confirmation_no']}.pdf", "wb"
> >                     ) as appSlipPdf:
> >                         appSlipPdf.write(appslip.content)
> >                     if os.path.exists(
> >                         f"{resp.json()['appointment_confirmation_no']}.pdf"
> >                     ):
> >                         print(
> >                             "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
> >                         )
> >                     else:
> >                         print("\nAppointment Slip Download Failed...")
> > 
> >                 except Exception as e:
> >                     print(str(e))
> >                     print("\n\n")
> > ```
> > 
> 
> This is same as before right?
> 
> > **OR**
> > 
> > ```
> > try:
> >                     appSlipBase = (
> >                         APPOINTMENT_SLIP_URL
> >                         + f"?appointment_id={resp.json()['appointment_id']}"
> >                     )
> >                     appslip = requests.get(appSlipBase, headers=request_header)
> >                     with open(
> >                         f"{resp.json()['appointment_id']}.pdf", "wb"
> >                     ) as appSlipPdf:
> >                         appSlipPdf.write(appslip.content)
> >                     if os.path.exists(
> >                         f"{resp.json()['appointment_id']}.pdf"
> >                     ):
> >                         print(
> >                             "\nDownload Successful. Check the Current Working Directory for the Appointment Slip."
> >                         )
> >                     else:
> >                         print("\nAppointment Slip Download Failed...")
> > 
> >                 except Exception as e:
> >                     print(str(e))
> >                     print("\n\n")
> > ```
> > 
> 
> This doesn't work
> image
> 
> 

No, the first one is not the same. The symbol in URL `?` got misplaced by `&`. I've solved it and as you've shown the second one is not working then the first snippet will work. There is a change so please reflect in in your local repo.

**Comment by dhhruv** [comment]
@Nakul93 Let me know the output when done.

**Comment by Nakul93** [comment]
Ok now it works. Thanks @dhhruv

**Comment by dhhruv** [comment]
Sure. Will change it afterwards.

**Comment by dhhruv** [comment]
Please check #477 for the same.

**Pull request closed without merging**